### DISCUSSION TITLE
Remix Vite is producing client bundles for resource routes

### DISCUSSION THREAD

**Discussion by danprince** [remix-run/remix#9401] [Q&A]
We're trying to migrate a moderately complex Remix app to use Vite but the main blocker we're running into is that there is server code in our client bundles (`process` not defined style errors) and that Rollup fails with `[commonjs--resolver] Server-only module referenced by client`.

Important to note that these are not problems in Remix v2.9.1 unless you switch to `remix vite:dev` and `remix vite:build`.

The offending routes are resource routes that form our API endpoints. For example:

```
routes/
  api.[fields.create].ts
```

This route is a resource route and has no `default` export because it's designed to only ever run at the server.

Each of these API endpoints export an object that describes the shape of the endpoint.

```ts
export const endpoint = defineEndpoint({
  name: "fields.create",

  input: z.strictObject({ ... }),

  output: z.strictObject({ ... }),
});

export const loader = async (...) => {
  // Parses the request/response according to schemas in `endpoint` above
}
```

The client side of the API uses the types of these exports to infer the types for the request/response.

```ts
const client = createFetchClient<
  | typeof import("~/routes/api.[fields.create]")
>();
```

For this to work, the module must export the endpoint shape. This becomes problematic when those Zod schemas are imported from `.server` files.

Technically we could hack around this by exporting the `typeof` the endpoint shape, or wrapping every endpoint definition in `serverOnly$` macros but I think the bundling behaviour has changed for resource routes, which seems like a more important issue to address.

A resource route won't ever run at the client side, right? So routes without default exports shouldn't ever be considered for client side bundling.

We can probably work around this by moving our API routes away from Remix, but the resource route model has worked so well up to this point, that it would be a shame to lose it all.

Would be great to get a clarification as to whether this is the intended behaviour and it's a known edge case when migrating from Remix Classic to Remix Vite, or whether this is something specific to our app.

**Comment by kiliman** [comment]
It's not that resource routes will ever _run_ on the client. It's what you export. Anything you export from a route module, except for `loader`, `action`, and `headers` is available in both server and client bundles.

**Reply by danprince** [reply]
Thanks @kiliman, I'm still pretty confused though. Why would the exports of resource routes ever be in the client bundle at all (unless the client side of the app imports a resource route explicitly—we don't)? And why did this work fine pre-Vite?
